tensorboard/plugins/pr_curve/__init__.py

```diff
@@ -0,0 +1,195 @@
+
+  Returns:
+    A tensor summary used by the pr_curve plugin.
+  """
+  weights = weights if weights is not None else 1.0
+  dtype = predictions.dtype
+
+  scope = os.path.join('PRSummaries', tag)
+  with tf.name_scope(scope, values=[labels, predictions]):
+    # Cast labels to bool then back to float to ensure we have 0.0 or 1.0.
+    f_labels = tf.cast(labels, dtype)
+    # Ensure predictions are all in range [0.0, 1.0].
+    predictions = tf.minimum(1.0, tf.maximum(0.0, predictions))
+    # Get weighted true/false labels.
+    true_labels = f_labels * weights
+    false_labels = (1.0 - f_labels) * weights
+
+    # Before we begin, reshape everything to (num_values, num_classes).
+    # We have to do this after weights multiplication since weights broadcast.
+    shape = (-1, num_classes)
+    predictions = tf.reshape(predictions, shape)
+    true_labels = tf.reshape(true_labels, shape)
+    false_labels = tf.reshape(false_labels, shape)
+
+    # To compute TP/FP/TN/FN, we are measuring a classifier
+    #   C(t) = (predictions >= t)
+    # at each threshold 't'. So we have
+    #   TP(t) = sum( C(t) * true_labels )
+    #   FP(t) = sum( C(t) * false_labels )
+    #
+    # But, computing C(t) requires computation for each t. To make it fast,
+    # observe that C(t) is a cumulative integral, and so if we have
+    #   thresholds = [t_0, ..., t_{n-1}];  t_0 < ... < t_{n-1}
+    # where n = num_thresholds, and if we can compute the bucket function
+    #   B(i) = Sum( (predictions == t), t_i <= t < t{i+1} )
+    # then we get
+    #   C(t_i) = sum( B(j), j >= i )
+    # which is the reversed cumulative sum in tf.cumsum().
+    #
+    # We can compute B(i) efficiently by taking advantage of the fact that
+    # our thresholds are evenly distributed, in that
+    #   width = 1.0 / (num_thresholds - 1)
+    #   thresholds = [0.0, 1*width, 2*width, 3*width, ..., 1.0]
+    # Given a prediction value p, we can map it to its bucket by
+    #   bucket_index(p) = floor( p * (num_thresholds - 1) )
+    # so we can use tf.scatter_add() to update the buckets in one pass.
+
+    # First compute the bucket indices for each prediction value.
+    bucket_indices = (
+        tf.cast(tf.floor(predictions * (num_thresholds-1)), tf.int32))
+    # Adjust indicies by classes. For performance and simplicity, we keep
+    # the buckets (see below) as 1D array representing the tp/fp buckets for
+    # a (num_classes, num_thresholds) tensors.
+    # So, for each index in bucket_indices, its real index into this 1D array is
+    #   index + (num_thresholds * class_index)
+    class_indices = tf.constant(range(num_classes), shape=(1, num_classes))
+    bucket_indices += num_thresholds * class_indices
+
+    with tf.name_scope('variables'):
+      # Now create the variables which correspond to the bucket values.
+      # These are flat arrays with num_thresholds values per class.
+      tp_buckets_v = tf.contrib.framework.local_variable(
```

Comment:
Would it be possible to not use `tf.contrib`? It's pretty expensive to dereference that property, because AFAIK, it causes everything in contrib to be loaded into memory.